openhands/sdk/utils/truncate.py

```diff
@@ -10,35 +14,76 @@
 )
 
 
+def _save_full_content(content: str, save_dir: str, tool_prefix: str) -> str:
+    """Save full content to the specified directory and return the file path."""
+
+    save_dir_path = Path(save_dir)
+    save_dir_path.mkdir(exist_ok=True)
+
+    # Generate a unique filename with timestamp
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S_%f")[
+        :-3
+    ]  # microseconds to milliseconds
+    filename = f"{tool_prefix}_output_{timestamp}.txt"
+    file_path = save_dir_path / filename
+
+    try:
+        file_path.write_text(content, encoding="utf-8")
+        return str(file_path)
+    except Exception as e:
+        # If saving fails, return a descriptive error message
+        return f"<failed to save: {e}>"
+
+
 def maybe_truncate(
     content: str,
     truncate_after: int | None = None,
     truncate_notice: str = DEFAULT_TRUNCATE_NOTICE,
+    save_dir: str | None = None,
+    tool_prefix: str = "output",
 ) -> str:
     """
     Truncate the middle of content if it exceeds the specified length.
 
     Keeps the head and tail of the content to preserve context at both ends.
+    Optionally saves the full content to a file for later investigation.
 
     Args:
         content: The text content to potentially truncate
         truncate_after: Maximum length before truncation. If None, no truncation occurs
         truncate_notice: Notice to insert in the middle when content is truncated
+        save_dir: Working directory to save full content file in
+        tool_prefix: Prefix for the saved file (e.g., "bash", "browser", "editor")
 
     Returns:
         Original content if under limit, or truncated content with head and tail
-        preserved
+        preserved and reference to saved file if applicable
     """
     if not truncate_after or len(content) <= truncate_after or truncate_after < 0:
         return content
 
+    # Save full content if requested
+    saved_file_path = None
+    if save_dir:
+        saved_file_path = _save_full_content(content, save_dir, tool_prefix)
+
+    # Create enhanced truncation notice with file reference if available
+    enhanced_notice = truncate_notice
+    if saved_file_path and not saved_file_path.startswith("<failed"):
+        enhanced_notice = (
+            f"<response clipped><NOTE>Due to the max output limit, only part of the "
+            f"full response has been shown to you. The complete output has been "
+            f"saved to {saved_file_path} - you can use the str_replace_editor tool "
```

Comment:
```suggestion
            f"saved to {saved_file_path} - you can use other tool "
```

In cases where str replace editor does not exist